### PR TITLE
[M1522] Triggers to update revision tables

### PR DESCRIPTION
**Pull request by gridcell**

### CODE REVIEW

**Comment by gridcell** [comment]
[Create triggers for creating/updating/deleting records from revision tables](https://trello.com/c/kdhcGHhR/1522-create-triggers-for-creating-updating-deleting-records-from-revision-tables)

File: src/api/models/revisions.py [hunk 1]
```diff
@@ -0,0 +1,413 @@
...
+        FROM pg_index, pg_class, pg_attribute, pg_namespace
+        WHERE
+            pg_class.oid = table_name::regclass AND
+            indrelid = pg_class.oid AND
+            nspname = 'public' AND
+            pg_class.relnamespace = pg_namespace.oid AND
+            pg_attribute.attrelid = pg_class.oid AND
+            pg_attribute.attnum = any(pg_index.indkey)
+            AND indisprimary;
+
```
Comment (saanobhaai) [inline src/api/models/revisions.py:57]: could maybe lose some `\n`s

File: src/api/models/revisions.py [hunk 2]
```diff
@@ -0,0 +1,413 @@
...
+            "last_rev_id",
+            "table_name",
+            "updated_on",
+            "project_id"
+        )
+        VALUES (
+            rev_id_val,
+            TG_TABLE_NAME,
+            updated_on_val,
+            Case WHEN project_id_val is null THEN
```
Comment (saanobhaai) [inline src/api/models/revisions.py:134]: all-cap `CASE`?

File: src/api/models/revisions.py [hunk 3]
```diff
@@ -0,0 +1,413 @@
...
+        OR DELETE ON "api_collectrecord" FOR EACH ROW EXECUTE FUNCTION record_change();
+
+    CREATE TRIGGER benthic_attribute_trigger
+    AFTER
+    INSERT
+        OR
+    UPDATE
+        OR DELETE ON "benthic_attribute" FOR EACH ROW EXECUTE FUNCTION record_change();
+
+    /*
```
Comment (saanobhaai) [inline src/api/models/revisions.py:182]: Need commented one?

File: src/api/models/revisions.py [hunk 4]
```diff
@@ -0,0 +1,413 @@
+import uuid
+
+from django.contrib.gis.db import models
+
+
+class RecordRevision(models.Model):
+    rev_id = models.UUIDField(primary_key=True, default=uuid.uuid4, editable=False)
```
[the remark this answers is not in the document]
Comment (gridcell) [reply src/api/models/revisions.py:7]: Do you think there would be a need or advantage to having FKs?

**Comment by saanobhaai** [comment]
Will `dbbackup` and `dbrestore` include the trigger functions?

File: src/api/tests/test_revision_triggers.py [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+    assert qry.count() == 0
+
+    collect_record.delete()
+
+    qry = RecordRevision.objects.filter(record_id=pk, deleted=True)
+
+    assert qry.count() == 1
+
+
+def test_create_update_delete_record(db_setup, project1):
```
Comment (saanobhaai) [inline src/api/tests/test_revision_triggers.py:87]: Do we need both this and the previous test?

File: src/api/models/revisions.py [hunk 6]
```diff
@@ -0,0 +1,410 @@
...
+        )
+        ON CONFLICT DO NOTHING;
+
+        RETURN NULL;
+
+    END;
+    $$;
+
+    CREATE TRIGGER fish_species_trigger
+    AFTER
```
Comment (saanobhaai) [inline src/api/models/revisions.py:154]: Still feels like a lot of extra lines...

File: src/api/models/revisions.py [hunk 7]
```diff
@@ -0,0 +1,410 @@
...
+    class Meta:
+        db_table = "record_revision"
+        unique_together = ("table_name", "record_id")
+
+    def __str__(self):
+        return f"[{self.rev_id} {self.updated_on}] {self.table_name} {self.record_id}"
+
+class TableRevision(models.Model):
+    last_revision = models.OneToOneField("RecordRevision", on_delete=models.CASCADE)
+
```
Comment (saanobhaai) [inline src/api/models/revisions.py:24]: extra line

File: src/api/models/revisions.py [hunk 8]
```diff
@@ -0,0 +1,410 @@
+import uuid
+
+from django.contrib.gis.db import models
+
+
+class RecordRevision(models.Model):
+    rev_id = models.UUIDField(editable=False, unique=True, db_index=True)
```
Comment (saanobhaai) [inline src/api/models/revisions.py:7]: So, not a primary key?

File: src/api/models/revisions.py [hunk 9]
```diff
@@ -0,0 +1,410 @@
...
+    table_name = models.CharField(max_length=50, db_index=True, editable=False)
+    record_id = models.UUIDField(db_index=True, editable=False)
+    project_id = models.UUIDField(null=True, db_index=True, editable=False)
+    profile_id = models.UUIDField(null=True, editable=False)
```
Comment (saanobhaai) [inline src/api/models/revisions.py:11]: `db_index=True`?

File: src/api/models/revisions.py [hunk 10]
```diff
@@ -0,0 +1,410 @@
...
+    END;
+    $$
+    LANGUAGE plpgsql;
+
+    CREATE OR REPLACE FUNCTION record_change()
+    RETURNS TRIGGER LANGUAGE PLPGSQL AS $$
+
+    DECLARE
+        pk uuid;
+        record_revision_id int;
```
Comment (saanobhaai) [inline src/api/models/revisions.py:65]: Isn't this a `uuid`?

File: src/api/models/revisions.py [hunk 11]
```diff
@@ -0,0 +1,410 @@
...
+
+        /*
+          Update table revsions
+        */
+
+        INSERT INTO table_revision(
+            "last_revision_id"
+        )
+        VALUES (
+            record_revision_id
```
Comment (saanobhaai) [inline src/api/models/revisions.py:144]: Do we need `record_revision_id` at all? Can't we just use `rev_id_val` here, and remove the previous `SELECT`?

**Review by saanobhaai — CHANGES_REQUESTED** [review]
Looking hugely streamlined, much better

File: src/api/tests/conftest.py [hunk 12]
```diff
@@ -1,7 +1,9 @@
 import pytest
 from django.db import connection
 
-from api.models.view_models import model_view_migrations, SummarySiteViewModel
+from api.models.view_models import model_view_migrations
+from api.models import revisions
+from api.models.view_models.summary_site import SummarySiteViewModel
```
[the remark this answers is not in the document]
Comment (gridcell) [reply src/api/tests/conftest.py:6]: shouldn't even exist, removing

File: src/api/models/revisions.py (hunk 6 above)
Comment (gridcell) [reply src/api/models/revisions.py:154]: don't blame me, blame the sql formatter :)

**Pull request closed without merging**